### PR TITLE
 new: dev: #124: report liveness metric for BookKeeper daemon

### PR DESCRIPTION
**Pull request by jordanw-bq**
- Changes have been tested on a Qubole cluster

### CODE REVIEW

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java [hunk 1]
```diff
@@ -118,4 +127,40 @@ public static boolean isServerUp()
 
     return false;
   }
+
+  /**
+   * Starts a new thread to report the liveness metric at a fixed rate.
+   */
+  private static void scheduleLivenessMetric()
```
Comment (vrajat) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: I dont think you do not need these classes or the scheduled executor. You only need to register the gauge in the constructor. The statsdreporter (or any reporter) will call a get value regularly. You've already implemented the getValue function to return 1. Also the fact that you had to put the null condition in `registerLivenessCheck` is also a clue that you want to do something only once. 

So the change required is to move the gauge registration (Lines 155-162) to `startServer`

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java [hunk 2]
```diff
@@ -118,4 +127,40 @@ public static boolean isServerUp()
...
+
+    private void registerLivenessCheck()
+    {
+      if (metrics.getGauges().get(METRIC_BOOKKEEPER_LIVENESS_CHECK) == null) {
+        metrics.register(METRIC_BOOKKEEPER_LIVENESS_CHECK, new Gauge<Integer>()
+        {
+          @Override
+          public Integer getValue()
+          {
+            return 1;
```
Comment (vrajat) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: This is sufficient for now. I want to have a health registry for all components in the bookkeeper like the server and disk monitor. This function should check the health of all the components and then return 1. I'll file an issue for this.

**Comment by jordanw-bq** [comment]
@vrajat This implementation was intended just for registering the metric. We have another task for reporting that to Datadog via the StatsDReporter (QUB-47)

**Comment by vrajat** [comment]
Ack on the StatsdReporter. Can you also comment on the need for scheduler executor service ?

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 1 above)
Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: Carry-over from initially implementing it as a counter. I agree with your notes, and will update the gauge resgistration.

Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: Gauge registration has been updated.

**Pull request merged**